src/model_io/urdf/src/URDFModelExport.cpp

```diff
@@ -241,6 +242,68 @@ bool exportJoint(IJointConstPtr joint, LinkConstPtr parentLink, LinkConstPtr chi
     return ok;
 }
 
+/*
+ * Add a <link> and <joint> URDF elements to the specified parent element for the specified additional frame.
+ *
+ * URDF does not have the concept of frame (see https://discourse.ros.org/t/urdf-ng-link-and-frame-concepts/56),
+ * so for each additional frame we need to add a URDF "fake" mass-less link.
+ *
+ * If parent_element contains a pointer to a tag <parent_element></parent_element>, this function modifies it to be something like:
+ *
+ * <parent_element>
+ *   <link name="<frame_name>"/>
+ *   <joint name="<frame_name>_fixed_joint" type="fixed">
+ *      <origin xyz="0.1 0.2 0.3" rpy="0 0 3.1416"/>
+ *      <parent link="<link_name>"/>
+ *      <child link="<frame_name>"/>
+ *   </joint>
+ * <parent_element/>
+ *
+ * where the actual values of the added element and attributes depend on the input frame_name, link_H_frame, parent_link_name and direction_options arguments.
+ */
+enum exportAdditionalFrameDirectionOption {
+    FAKE_LINK_IS_CHILD,
+    FAKE_LINK_IS_PARENT
+};
+bool exportAdditionalFrame(const std::string frame_name, Transform link_H_frame, const std::string link_name, exportAdditionalFrameDirectionOption direction_option, xmlNodePtr parent_element)
+{
+    bool ok=true;
+
+    // Export fake link
+    xmlNodePtr link_xml = xmlNewChild(parent_element, NULL, BAD_CAST "link", NULL);
+    xmlNewProp(link_xml, BAD_CAST "name", BAD_CAST frame_name.c_str());
+
+    // Export fake joint
+    xmlNodePtr joint_xml = xmlNewChild(parent_element, NULL, BAD_CAST "joint", NULL);
+    std::string fake_joint_name = frame_name + "_fixed_joint";
+    xmlNewProp(joint_xml, BAD_CAST "name", BAD_CAST fake_joint_name.c_str());
+    xmlNewProp(joint_xml, BAD_CAST "type", BAD_CAST "fixed");
+
+    // origin
+    exportTransform(link_H_frame, joint_xml);
+
+    std::string parent_of_fake_joint, child_of_fake_joint;
+
+    if (direction_option == FAKE_LINK_IS_CHILD) {
+        parent_of_fake_joint = link_name;
+        child_of_fake_joint = frame_name;
+    } else {
+        // direction_option == FAKE_LINK_IS_PARENT
+        parent_of_fake_joint = frame_name;
+        child_of_fake_joint = link_name;
+    }
+
+    // parent
+    xmlNodePtr parent_xml = xmlNewChild(joint_xml, NULL, BAD_CAST "parent", NULL);
+    xmlNewProp(parent_xml, BAD_CAST "link", BAD_CAST parent_of_fake_joint.c_str());
+
+    // child
+    xmlNodePtr child_xml = xmlNewChild(joint_xml, NULL, BAD_CAST "child", NULL);
+    xmlNewProp(child_xml, BAD_CAST "link", BAD_CAST child_of_fake_joint.c_str());
+
```

Comment:
I see a common trend of using the method `xmlNewChild` and `xmlNewProp` sequentially. Can we not make a function that does the composition of these two steps?

Reply:
I see this also depends on how many properties we add in sequence. So the complexity of having a single function to handle all properties might be big.

Reply:
Yes, this could make sense, but at least initially I preferred to avoid start writing  a C++ wrapper on the top of the libxml2 API, without having a clear idea about our need from the libxml2's API. Furthermore, I am a bit afraid that adding a layer between the libxml2's API and the parser could impact readability. However, a C++ wrapper could make sense, but I would prefer to complete the URDF exporter implementation to understand  our needs, and eventually do an libxml2 abstraction as a refactoring. What do you think?

Reply:
Agreed.